src/request/export/ExportWords.js

```diff
@@ -21,94 +22,112 @@ class ExportWords extends Nimiq.Observable {
-
-        $privacyWarningButton.addEventListener('click', this._goToShowWords.bind(this));
-        $recoveryWordsButton.addEventListener('click', this._goToValidateWords.bind(this));
-        this._privacyWarningPassphraseBox.on(PassphraseBox.Events.SUBMIT, this._passphraseSubmitted.bind(this));
-        this._validateWords.on(ValidateWords.Events.VALIDATED, this._finish.bind(this));
-        this._validateWords.on(ValidateWords.Events.BACK, this._goToShowWords.bind(this));
-        this._validateWords.on(ValidateWords.Events.SKIP, this._finish.bind(this));
+        /* eslint-disable no-new */
+        new ProgressIndicator(this._$noRecovery.querySelector('.progress-indicator'), 3, 1);
+        new ProgressIndicator($recoveryWordsPage.querySelector('.progress-indicator'), 3, 2);
+        new ProgressIndicator($validateWordsPage.querySelector('.progress-indicator'), 3, 3);
+        /* eslint-enable no-new */
+
+        // events
+        $noRecoverySkip.addEventListener('click', event => {
+            event.preventDefault();
+            this._resolve({ success: false });
+        });
+        $recoveryWordsSkip.addEventListener('click', event => {
+            event.preventDefault();
+            this._resolve({ success: true });
+        });
+        this._wordsPasswordBox.on(PassphraseBox.Events.SUBMIT, this._passphraseSubmitted.bind(this));
+        $recoveryWordsContinue.addEventListener('click', () => {
+            this._validateWords.reset();
+            window.location.hash = ExportWords.Pages.VALIDATE_WORDS;
+        });
+        this._validateWords.on(ValidateWords.Events.VALIDATED, () => this._resolve({ success: true }));
+        this._validateWords.on(ValidateWords.Events.SKIP, () => this._resolve({ success: true }));
     }
 
     run() {
-        this._privacyWarningPassphraseBox.reset();
-        window.location.hash = ExportWords.Pages.PRIVACY;
+        this._wordsPasswordBox.reset();
+        window.location.hash = ExportWords.Pages.RECOVERY_WORDS_INTRO;
         if (TopLevelApi.getDocumentWidth() > Constants.MIN_WIDTH_FOR_AUTOFOCUS) {
-            this._privacyWarningPassphraseBox.focus();
+            this._wordsPasswordBox.focus();
         }
     }
 
     /**
-     * @param {string} phrase
+     * @param {string} password
      */
-    async _passphraseSubmitted(phrase) {
+    async _passphraseSubmitted(password) {
         TopLevelApi.setLoading(true);
-        const passphraseBuffer = phrase ? Utf8Tools.stringToUtf8ByteArray(phrase) : undefined;
+
+        let passwordBuffer;
+        if (this._password) {
+            passwordBuffer = this._password;
+        } else if (password) {
+            passwordBuffer = Utf8Tools.stringToUtf8ByteArray(password);
+        }
+
         /** @type {Key?} */
         let key = null;
+
         try {
-            key = await KeyStore.instance.get(this._request.keyInfo.id, passphraseBuffer);
+            key = await KeyStore.instance.get(this._request.keyInfo.id, passwordBuffer);
```

Comment:
If we already know the `_password`, then we most likely also already know the `this._key`. I think getting it from the database and running the decryption _again_ is unecessary when `this._key` is already available.